src/vs/platform/contextkey/common/contextkey.ts

```diff
@@ -241,6 +269,370 @@ export abstract class ContextKeyExpr {
+		} else {
+			throw this._errExpectedButGot(`'true', 'false', '(', KEY, KEY '=~' regex, KEY [ ('==' | '!=' | '<' | '<=' | '>' | '>=' | 'in' | 'not' 'in') value ]`, this._peek());
+		}
+	}
+
+	private _value(): string { // TODO@ulugbekna: match all at once and then `switch` on the type
+		if (this._match(TokenType.Str, TokenType.QuotedStr)) {
+			return this._previous().lexeme!;
+		} if (this._match(TokenType.True)) {
+			return 'true';
+		} if (this._match(TokenType.False)) {
+			return 'false';
+		} if (this._match(TokenType.In)) { // we support `in` as a value, e.g., "when": "languageId == in" - exists in existing extensions
+			return 'in';
+		} else {
+			return ''; // this allows "when": "foo == " which's used by existing extensions
+		}
+	}
+
+	// careful: this can throw if current token is the initial one (ie index = 0)
+	private _previous() {
+		return this._tokens[this._current - 1];
+	}
+
+	private _match(...tokens: TokenType[]) {
+		for (const token of tokens) {
+			if (this._check(token)) {
+				this._advance();
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	private _advance() {
+		if (!this._isAtEnd()) {
+			this._current++;
+		}
+		return this._previous();
+	}
+
+	private _consume(type: TokenType, message: string) {
+		if (this._check(type)) {
+			return this._advance();
+		}
+
+		throw this._errExpectedButGot(message, this._peek());
+	}
+
+	private _errExpectedButGot(expected: string, got: Token) {
+		return this._error(`Expected ${expected} but got '${this._nameToken(got)}' at offset ${got.offset}.`);
+	}
+
+	private _errUnexpected(token: Token) {
+		return this._error(`Unexpected '${this._nameToken(token)}' at offset ${token.offset}.`);
+	}
+
+	// TODO@ulugbekna: the whole error reporting needs reworking - especially, when we introduce it to package.json linting
+	private _error(errMsg: string) {
+		this._parsingErrors.push(errMsg);
+		return new ParseError();
+	}
+
+	private _nameToken(token: Token): string {
+		if (token.lexeme) { return token.lexeme!; }
+
+		switch (token.type) {
+			case TokenType.LParen:
+				return '(';
+			case TokenType.RParen:
+				return ')';
+			case TokenType.Neg:
+				return '!';
+			case TokenType.Eq:
+				return '==';
+			case TokenType.NotEq:
+				return '!=';
+			case TokenType.Lt:
+				return '<';
+			case TokenType.LtEq:
+				return '<=';
+			case TokenType.Gt:
+				return '>=';
+			case TokenType.GtEq:
+				return '>=';
+			case TokenType.RegexOp:
+				return '=~';
+			case TokenType.True:
+				return 'true';
+			case TokenType.False:
+				return 'false';
+			case TokenType.In:
+				return 'in';
+			case TokenType.Not:
+				return 'not';
+			case TokenType.And:
+				return '&&';
+			case TokenType.Or:
+				return '||';
+			case TokenType.EOF:
+				return 'EOF';
+			default:
+				throw illegalState(`all other tokens must have a lexeme : ${JSON.stringify(token)}`);
+		}
+	}
+
+	private _check(type: TokenType) {
+		return !this._isAtEnd() && this._peek().type === type;
```

Comment:
The check for `!this._isAtEnd()` doesn't seem necessary here

Reply:
Great catch!